aten/src/ATen/native/Copy.cpp

```diff
@@ -26,6 +27,53 @@ namespace {
 
 using namespace at;
 
+bool fbgemm_copy_transpose_valid(const Tensor& self, const Tensor& src) {
+  const int MIN_SZ = 16 * 32;
+  if ((self.device().is_cpu() && src.device().is_cpu()) &&
+      (self.layout() == c10::kStrided) && (src.layout() == c10::kStrided) &&
+      !self.is_sparse() && !src.is_sparse() && self.is_contiguous() &&
+      (self.is_conj() == src.is_conj()) && (self.is_neg() == src.is_neg()) &&
+      !self.is_complex() && !src.is_complex() &&
+      self.sizes().equals(src.sizes()) && self.dim() >= 2 &&
+      src.size(src.dim() - 1) * src.size(src.dim() - 2) >= MIN_SZ &&
+      src.stride(src.dim() - 2) == 1 && src.stride(src.dim() - 1) == src.size(src.dim() - 2) &&
+      !(src.size(src.dim() - 2) == 1 && src.size(src.dim() - 1) == 1)) {
+      // Check src is in contiguous block
+      for (long i = 0; i < src.dim() - 2; i++) {
+        if (!(src.stride(i) == ((i + 1) == (src.dim() - 2)) ?
+                src.stride(src.dim() - 1) * src.size(src.dim() - 1) :  src.stride(i + 1) * src.size(i + 1))){
+              return false;
+            }
+      }
+  } else {
+    return false;
+  }
+  return true;
+}
+
+void fbgemm_copy_transpose_same_type(Tensor& self, const Tensor& src) {
+  auto block_size = src.size(src.dim() - 1) * src.size(src.dim() - 2);
+  auto ntrans = src.numel() / block_size;
+  AT_DISPATCH_ALL_TYPES_AND(kBFloat16, src.scalar_type(),
+    "fbgemm_transpose_copy_same_type", [&] {
+    at::parallel_for(
```

Comment:
It would be very helpful if you can describe the decomposition heuristics and parallel algorithm.  How the last 2D subtensor is blocked, which dimension is being parallelized, and how to handle the edge case?